### PR TITLE
Make PromQL metrics un-global, remove rules dep on promql.Engine

### PR DESCRIPTION
**Pull request by fabxc**

### CODE REVIEW

File: template/template.go [hunk 1]
```diff
@@ -59,34 +58,13 @@ func (q queryResultByLabelSorter) Swap(i, j int) {
 	q.results[i], q.results[j] = q.results[j], q.results[i]
 }
 
-func query(ctx context.Context, q string, ts time.Time, queryEngine *promql.Engine) (queryResult, error) {
-	query, err := queryEngine.NewInstantQuery(q, ts)
+type QueryFunc func(context.Context, string, time.Time) (promql.Vector, error)
+
+func query(ctx context.Context, q string, ts time.Time, query QueryFunc) (queryResult, error) {
+	vector, err := query(ctx, q, ts)
```
Comment (Bplotka) [inline template/template.go]: can we rename `query` to `queryFn` or `queryFunc`? Looks like recursive call at first glance (:

File: rules/manager.go [hunk 2]
```diff
@@ -107,12 +107,42 @@ const (
 	ruleTypeRecording = "recording"
 )
 
+// QueryFunc processes PromQL queries.
+type QueryFunc func(ctx context.Context, q string, t time.Time) (promql.Vector, error)
+
+// QueryFuncFromEngine returns a new query function that executes instant queries against
+// the given engine.
+// It converts scaler into vector results.
+func QueryFuncFromEngine(engine *promql.Engine) QueryFunc {
```
Comment (Bplotka) [inline rules/manager.go]: maybe just `EngineQueryFunc` instead of longer `QueryFuncFromEngine`?

File: rules/alerting.go [hunk 3]
```diff
@@ -83,7 +83,9 @@ type Alert struct {
 	Value float64
 	// The interval during which the condition of this alert held true.
 	// ResolvedAt will be 0 to indicate a still active alert.
-	ActiveAt, ResolvedAt time.Time
+	ActiveAt   time.Time
+	FiringAt   time.Time
```
Comment (brancz) [inline rules/alerting.go]: Maybe `LastFiredAt`? At first glance I wasn't sure what this was going to be.

Comment (fabxc) [reply rules/alerting.go]: Mh, it's not the last fired at timestamp though but the timestamp at which it first started firing.

Comment (brancz) [reply rules/alerting.go]: Right, then `FirstFiredAt`?

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -233,8 +237,8 @@ func main() {
 
 	ruleManager := rules.NewManager(&rules.ManagerOptions{
 		Appendable:  fanoutStorage,
-		Notifier:    notifier,
-		QueryEngine: queryEngine,
+		Query:       rules.EngineQueryFunc(queryEngine),
```
Comment (Gouthamve) [inline cmd/prometheus/main.go]: Could this be named `QueryFunc` and `NotifyFunc`?

File: cmd/prometheus/main.go [hunk 5]
```diff
@@ -96,6 +96,9 @@ func main() {
 		notifier: notifier.Options{
 			Registerer: prometheus.DefaultRegisterer,
 		},
+		queryEngine: promql.EngineOptions{
+			Metrics: prometheus.DefaultRegisterer,
```
[the remark this answers is not in the document]
Comment (fabxc) [reply cmd/prometheus/main.go]: Yea, neither is great I guess. But `Metrics` is at least a bit less generic.

**Pull request merged**